### PR TITLE
major size reduction

### PR DESCRIPTION
**Pull request by peterferrie**

### CODE REVIEW

**Comment by faissaloo** [comment]
I won't be merging this for two reasons:  
1. It's slower  
2. It's overcomplicated  

I'll explain:  
1. Accessing memory is slow, the PUSHing on line 54 accesses memory and is therefore slow.  
2. Instructions that do more than one thing are usually slower than their simple alternatives and are usually leftovers from early versions of x86 (pre-P4). Stuff like BSWAP, SCASD and STOSB should all be avoided.  
3. You're using too many labels, it makes the code hard to follow.  

I'll leave this open for like a day in case there's anything else you want to say, otherwise I'll just close it.

**Comment by peterferrie** [comment]
> I won't be merging this for two reasons:
> 1. It's slower

No, it's much faster than what you have right now.

> 1. It's overcomplicated
> 
> I'll explain:
> 1. Accessing memory is slow, the PUSHing on line 54 accesses memory and is
>    therefore slow.

Scanning the same string _repeatedly_ is slow.  You have an
exponential increase in search size.  I resume from the replacement
point.  My push instruction is executed once, and my string search is
faster overall, the "penalty" is more than compensated.

> 1. Instructions that do more than one thing are usually slower than their
>    simple alternatives and are usually leftovers from early versions of x86
>    (pre-P4). Stuff like BSWAP, SCASD and STOSB should all be avoided.

The BSWAP executes exactly once.  The faster string search more than
compensates here, too.  String opcodes have their place.  Avoiding
them entirely is an extreme position.

> 1. You're using too many labels, it makes the code hard to follow.

I've introduced two labels which make the flow more linear.

> I'll leave this open for like a day in case there's anything else you want
> to say, otherwise I'll just close it.

If you're going to highlight the small size of the replacement, then
why object to making it _really_ small?

**Comment by faissaloo** [comment]
I just benchmarked it and you were right, it is **considerably** faster. So I'll merge this now. Thanks alot for this and sorry for giving you crap.

**Comment by peterferrie** [comment]
> I just benchmarked it and you were, right, it is **considerably** faster. So
> I'll merge this now. Thanks alot for this and sorry for giving you crap.

No worries.  Thanks for considering it.
